sourcemap/es_store.go

```diff
@@ -0,0 +1,188 @@
+	sourcemapServiceNameStr    = "sourcemap.service.name"
+	sourcemapServiceVersionStr = "sourcemap.service.version"
+	sourcemapSourcemapStr      = "sourcemap.sourcemap"
+	sourceStr                  = "_source"
+	termStr                    = "term"
+	timestampStr               = "@timestamp"
+	valueStr                   = "value"
+
+	emptyResult = ""
+
+	errMsgParseSourcemap = "Could not parse Sourcemap."
+)
+
+var (
+	errMsgESFailure         = "failure querying ES"
+	errSourcemapWrongFormat = errors.New("Sourcemapping ES Result not in expected format")
+)
+
+type esStore struct {
+	client *elasticsearch.Client
+	index  string
+	logger *logp.Logger
+}
+
+type esSourcemapResponse struct {
+	Hits struct {
+		Total struct {
+			Value int
+		}
+		Hits []struct {
+			Source struct {
+				Sourcemap struct {
+					Sourcemap string
+				}
+			} `json:"_source"`
+		}
+	} `json:"hits"`
+}
+
+func (s *esStore) fetch(name, version, path string) (string, error) {
+	response, err := s.runSearchQuery(name, version, path)
+	if err != nil {
+		return "", errors.Wrap(err, errMsgESFailure)
+	}
+	defer response.Body.Close()
+	// handle error response
+	if response.IsError() {
+		if response.StatusCode == http.StatusNotFound {
+			return "", nil
+		}
+		b, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return "", errors.Wrap(err, errMsgParseSourcemap)
+		}
+		return "", errors.New(fmt.Sprintf("%s %s", errMsgParseSourcemap, b))
+	}
+
+	// parse response
+	return parse(response, name, version, path, s.logger)
+}
+
+func (s *esStore) runSearchQuery(name, version, path string) (*esapi.Response, error) {
+	// build and encode the query
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(query(name, version, path)); err != nil {
+		return nil, err
+	}
+
+	// Perform the runSearchQuery request.
+	return s.client.Search(
+		s.client.Search.WithContext(context.Background()),
+		s.client.Search.WithIndex(s.index),
+		s.client.Search.WithBody(&buf),
+		s.client.Search.WithTrackTotalHits(true),
+		s.client.Search.WithPretty(),
+	)
+}
+
+func parse(response *esapi.Response, name, version, path string, logger *logp.Logger) (string, error) {
+	var esSourcemapResponse esSourcemapResponse
+	if err := json.NewDecoder(response.Body).Decode(&esSourcemapResponse); err != nil {
+		return "", err
+	}
+	hits := esSourcemapResponse.Hits.Total.Value
+	if hits == 0 {
+		return emptyResult, nil
+	}
+
+	var esSourcemap string
+	if hits > 1 {
+		logger.Warnf("%d sourcemaps found for service %s version %s and file %s, using the most recent one",
+			hits, name, version, path)
+	}
+	esSourcemap = esSourcemapResponse.Hits.Hits[0].Source.Sourcemap.Sourcemap
+	// until https://github.com/golang/go/issues/19858 is resolved
+	if esSourcemap == emptyResult {
+		return emptyResult, errSourcemapWrongFormat
+	}
+	return esSourcemap, nil
+}
+
+func query(name, version, path string) map[string]interface{} {
+	return map[string]interface{}{
+		queryStr: map[string]interface{}{
```

Comment:
If you think the usage of const does make it worse in terms of readability, I can switch back to current behavior. The performance and amount of allocations is the same. 

I do not see the point in adding all those methods, as to me that makes it really hard to read, plus it adds slightly more allocations.

Reply:
> plus it adds slightly more allocations

I'll argue that that is not remotely important and that we should focus on the important stuff. Readability is important because we spend the whole day reading code. On that note (constants aside):

```golang
boolStr: map[string]interface{}{
    mustStr: []map[string]interface{}{
        {termStr: map[string]interface{}{key: value}},
	{termStr: map[string]interface{}{key: value}},
	{termStr: map[string]interface{}{key: value}},
	{boolStr: map[string]interface{}{
	     shouldStr: []map[string]interface{}{
	         {termStr: map[string]interface{}{key: map[string]interface{}{
		     value: path,
		     boost: 2.0,
		  }}},
		  {termStr: map[string]interface{}{key: map[string]interface{}{value}}}}}}},
```

```golang
boolean(
	must(
	    term(key, value),
	    term(key, value),
	    term(key, value),
	    term(key, value),
	    boolean(
	        should(
		    boostedTerm(key, value),
		    term(key, value))))))
```

I want to understand, as objectively as possible, what is it that makes it for you the second case harder to read than the first one. 
In the first case, ~46% of information is almost meaningless (noise), while the second one is closer to the actual JSON query, and ~39% shorter (by char count) to do the same thing. 
The fact that is so much closer to JSON (the thing that comes out of it and we are trying to reflect there) makes it way easier for me to interpret it, find the parameters, and see at a glance the shape of the query. 

I think I am the only one regularly working on a 13" display (and large font text), so my screen is filled up with all the `map[string]interface{}{` and I have a very hard time reading large nested maps, maybe that's why I am the only one complaining about this.

Reply:
>plus it adds slightly more allocations
>> I'll argue that that is not remotely important and that we should focus on the important stuff. Readability is important because we spend the whole day reading code. 

Just for clarification, I _was_ arguing for readability, you only quoted one part of my sentence. The purpose of bringing up the allocations was, that if we have different opinions on what is more readable, we should go with the approach that has less allocations.  

The current approach mirrors how the Elasticsearch query looks like, and to me that is easily readable. Your suggestion adds ~10 small methods that I need to look at to verify behavior, adding a bit of mental overhead. IMO your solution is more elegant, but the current one is easier to read. None of this is making a big difference, and I'd like to avoid spending more time on this discussion, therefore I am moving forward with your approach.